### PR TITLE
box: fix memory leaks on `ER_MULTISTATEMENT_TRANSACTION` in DDL

### LINKED ISSUE #8773: Memory leaks in memtx_space_build_index and memtx_space_check_format functions.
**Bug description**

The suspects do not free the allocated iterator if `txn_check_singlestatement` check is failed.

memtx_space_build_index:
- Iterator allocation: [memtx_space.c +1241](https://github.com/tarantool/tarantool/blob/d2d068245db3c2947fb3b47f628e3eb37325175b/src/box/memtx_space.c#L1241)
- Return without deallocation: [memtx_space.c +1254](https://github.com/tarantool/tarantool/blob/d2d068245db3c2947fb3b47f628e3eb37325175b/src/box/memtx_space.c#L1254)

memtx_space_check_format:
- Iterator allocation: [memtx_space.c +990](https://github.com/tarantool/tarantool/blob/d2d068245db3c2947fb3b47f628e3eb37325175b/src/box/memtx_space.c#L990)
- Return without deallocation: [memtx_space.c +995](https://github.com/tarantool/tarantool/blob/d2d068245db3c2947fb3b47f628e3eb37325175b/src/box/memtx_space.c#L995)

**Rerpducer**

```lua
box.cfg{memtx_use_mvcc_engine = false}

s = box.schema.space.create('s')
s:create_index('pk', {parts = {{1, 'scalar'}}})
s:create_index('sk', {unique = false, hint = true})

-- Fill spaces to invoke format check/index rebuild.
for i = 1, 10 do s:insert({i, i, i}) end

function memtx_space_build_index_leak()
    print(pcall(function()
        box.begin()
        -- The first index rebuild because of `hint` change.
        s.index.sk:alter({hint = false})
        -- The second index rebuild because of `unique` change.
        s.index.sk:alter({unique = true})
        box.commit()
    end))
    box.rollback()
end

function memtx_space_check_format_leak()
    s.index.sk:alter({parts = {{1, 'scalar'}}})
    print(pcall(function()
        box.begin()
        s:format({{'id', 'scalar'}})
        s:format({{'id', 'number'}})
        s:format({{'id', 'unsigned'}})
        box.commit()
    end))
    box.rollback()
end

memtx_space_check_format_leak()
memtx_space_build_index_leak()

os.exit()
```

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
Closes #8773

### CODE REVIEW

**Review by andreyaksenov — APPROVED** [review]

File: test/box-luatest/gh_8733_ddl_memory_leaks_test.lua [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        end
+    end)
+end)
+
+g.after_all(function(cg)
+    cg.server:drop()
+end)
+
+-- Checks that memtx space index build does not leak on multi-statement
+-- transaction error.
```
Comment (locker) [inline test/box-luatest/gh_8733_ddl_memory_leaks_test.lua:25]: How does this test check that there's no leak?

File: src/box/memtx_space.c [hunk 2]
```diff
@@ -991,8 +991,10 @@ memtx_space_check_format(struct space *space, struct tuple_format *format)
 	if (it == NULL)
 		return -1;
 
-	if (txn_check_singlestatement(txn, "space format check") != 0)
+	if (txn_check_singlestatement(txn, "space format check") != 0) {
+		iterator_delete(it);
 		return -1;
+	}
```
Comment (locker) [inline src/box/memtx_space.c:997]: Better move the check before the iterator construction.

Comment (CuriousGeorgiy) [reply src/box/memtx_space.c]: Fixed.

File: test/box-luatest/gh_8733_ddl_memory_leaks_test.lua (hunk 1 above)
Comment (CuriousGeorgiy) [reply test/box-luatest/gh_8733_ddl_memory_leaks_test.lua]: Dropped this test.

**Review by locker — APPROVED** [review]

**Comment by locker** [comment]
Wrong ticket. Please fix.

**Comment by CuriousGeorgiy** [comment]
@locker fixed ticket number.

**Comment by locker** [comment]
Cherry-picked to 2.11 and 2.10.

**Pull request merged**